### ISSUE TITLE
MSVC 2015 runtime question

### ISSUE THREAD

**Issue by Siecje** [takluyver/pynsist#84]
What needs the MSVC 2015 runtime? NSIS? Python? I wouldn't think it is needed to run the application since that is just Python and then the .pyw file.

Can't you just bundle the files that the Vista .msu installs? So when building the installer add the Vista 32-bit or 64-bit files to the installer?

Also both of the Vista links are 404. The other are fine.

```
https://cdn.rawgit.com/takluyver/pynsist/msvcrt-tag1/x64/Windows6-KB2999226-x64.msu
https://cdn.rawgit.com/takluyver/pynsist/msvcrt-tag1/x86/Windows6-KB2999226-x86.msu
```

**Comment by takluyver** [comment]
It's needed to run Python. Python on Windows is compiled against the Microsoft VC runtime.

I believe the runtime depends on the OS you're running, not the version of Python you use, so while it's possible to bundle them, you'd have to bundle all eight msu packages. Maybe I'm wrong about that...

Thanks for the heads up about the vista links, I'll fix that.

**Comment by takluyver** [comment]
Those links should work in master - it needed to be `Windows6.0` instead of `Windows6`

**Comment by Siecje** [comment]
Okay thanks. Python.org only has a 32-bit and 64-bit version. And I don't believe their installer has to download anything.

**Comment by takluyver** [comment]
I presume the Python.org installer bundles all of the MSVCRT versions it needs. We only need to install MSVCRT if we use the 'embeddable' builds of Python, not if we use the installer. But the 'embeddable' build is preferable for other reasons.

@zooba: if we embed 32-bit Python and run on a 64-bit system, which MSVCRT do we need? I've been assuming we use the x64 version to match Windows, but thinking about it, it would seem to make more sense that we'd need the x86 version.

**Comment by Siecje** [comment]
Wouldn't the files for Vista also work for the newer Windows versions?

You think the the python.org installers have the 4 different files bundled in the installer?
For example the 64-bit installer would have the required files for Vista, 7, 8, 8.1 and only install the correct files?

**Comment by Siecje** [comment]
What do you mean by

> We only need to install MSVCRT if we use the 'embeddable' builds of Python, not if we use the installer.

**Comment by takluyver** [comment]
From what I'm told, each package is specific to one version of Windows; we can't just install the Vista version on newer versions of Windows. The relevant MSVCRT is installed by default on Windows 10, so we don't need it there. I've just unpacked the Python 3.5.2 installer and it does appear to contain eight copies of the MSVCRT runtime.

There are two options for how Python is included in a Pynsist installer: 'bundled' or 'installer'. The 'bundled' format is only available from Python 3.5. We only need to deal with MSVCRT directly for 'bundled' Python, because the installer already includes it. See the docs for more info:

http://pynsist.readthedocs.io/en/latest/cfgfile.html#python-section

**Comment by zooba** [comment]
The main reason the installers are separate for each Windows version is because they are update packages, so they have different detection routines and metadata. The files inside should be identical, otherwise you wouldn't be able to do applocal installs.

You'll need the C runtime that matches your python.exe. So if it's 32-bit Python, you need 32-bit CRT. *However*, the 64-bit installer packages include both 32-bit and 64-bit versions, and because it's an OS update it needs to match the target OS architecture.

In short:
* 32-bit python.exe ==> 32-bit ucrtbase.dll
* 64-bit python.exe ==> 64-bit ucrtbase.dll
* Windows...-KB2999226-**x86**.msu ==> 32-bit Windows
* Windows...-KB2999226-**x64**.msu ==> 64-bit Windows

(And for the side question about the python.org installers - if you get the 1MB web installer then we only download a single update package if needed, but the 30MB executable installer includes all of the packages.)

**Comment by takluyver** [comment]
Thanks @zooba !

So using the msu packages, we'd need all 8 for a 32-bit application, but we could get away with just 4 for a 64-bit application, because it's not going to run on 32-bit Windows anyway. Right? Maybe it would be better to download them at build time and bundle them into the installers, putting up with 4-6MB extra.

> The files inside should be identical, otherwise you wouldn't be able to do applocal installs.

This 'applocal' install of which you speak... would that mean putting the DLLs alongside Python.exe in the install directory? What are the downsides of that? Is it just `ucrtbase.dll`, or other files too? If that's practical, it sounds simpler than interacting with Windows update. It would also be less added size than bundling the MSU packages.

**Comment by zooba** [comment]
The downsides are your application won't necessarily get security updates automatically, and you take the extra size hit even when you don't need it.

There's an updated note on [this blog post](https://blogs.msdn.microsoft.com/vcblog/2015/03/03/introducing-the-universal-crt/) with details on how to do applocal redistribution. At the very least, I'd suggest making it conditional against Windows 10 (or detecting an existing install, if that's easier), but otherwise it's fine just not recommended.

**Comment by takluyver** [comment]
Thanks, I think I'm going to look into doing that.

- We're currently checking `IfFileExists "$SYSDIR\ucrtbase.dll"` to decide if we need to install the CRT on Windows 8.1 and below. Will that check work on Windows 10 too, or should we check for that separately?
- How frequent are the security updates we'd be missing?

**Comment by zooba** [comment]
That check should be fine. Technically that file isn't part of the public surface area of the CRT, but I think it's exposed enough already that it can't reasonably be taken away.

I have no idea how frequent the updates have been, but for anything critical they'll be released nearly immediately and more minor fixes probably come out on Patch Tuesday or in general OS updates.

**Comment by takluyver** [comment]
After a bit of a debate with the SDK installer, I have what I think are the right files: for each of x86 and x64, there are 41 DLLs: `ucrtbase.dll` plus 40 `api-ms-win-*.dll`. They total a bit under 2MB (unpacked) for each. Does that sound right?

From the blog post you linked, I believe that if I all I have to do is put these 41 DLLs in the same folder as `Python.exe`. Sorry to check what's probably obvious - my Windows installation already has the universal CRT, so I can't easily test what happens without it.

I'll keep thinking about the security implications. One way would be to give application developers a choice: simple but potentially insecure applocal installation, versus the more error-prone but hopefully safer installation of MSU packages. But then I'd have to maintain both mechanisms, and most people will probably use whatever the default is anyway...

**Comment by takluyver** [comment]
I've gone the applocal route in PR #87. I'll leave it as a PR for a few days so people for anyone who wants to comment, but I'm inclined to merge it.

I'm also planning to make applocal Python the default for Python 3.6 and above.

**Comment by Siecje** [comment]
What does the applocal Python do besides include the MSVC .dll's in the installation directory?

Would another option be include all of the .msu's in the installer and then run one if required?

Personally I'm very happy with the PR 👍

**Comment by takluyver** [comment]
> What does the applocal Python do besides include the MSVC .dll's in the installation directory?

It checks whether the CRT is already installed on the system, and skips installing the dlls if so.

> Would another option be include all of the .msu's in the installer and then run one if required?

It would be another option, but it keeps much of the complexity of the current approach - you still have to check the Windows version and invoke a separate process to install the msus. You also have to ship 4 or 8 different msus for different platforms it may be installed on, so the installers would be bigger.

I've gone for the applocal approach because it's simpler, and it should make it easier to do per-user installs in the future. If you prefer to package the MSUs and install them as necessary, it should be possible to do that with a custom template, basing it on the code from 1.8.
